### PR TITLE
Add FFI functions to retrieve secret key protection info.

### PR DESCRIPTION
**Pull request by ni4**
This PR adds a bunch of FFI functions, which allows to get information about the key protection (and whether secret key is available at all, like cases where secret key is offloaded or stored on the smartcard).

Related issue #1217

### CODE REVIEW

File: src/lib/rnp.cpp [hunk 1]
```diff
@@ -6164,6 +6178,130 @@ try {
...
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    if (!pgp_key_has_encryption_info(key->sec)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+
+    if (key->sec->pkt.sec_protection.s2k.specifier == PGP_S2KS_ITERATED_AND_SALTED) {
+        *iterations = pgp_s2k_decode_iterations(key->sec->pkt.sec_protection.s2k.iterations);
+    } else {
+        *iterations = 1;
```
Comment (dewyatt) [inline src/lib/rnp.cpp:6299]: I think 0 or an error would make more sense.

Comment (ni4) [reply src/lib/rnp.cpp:6299]: @dewyatt here the only cases left for now are `PGP_S2KS_SIMPLE` and `PGP_S2KS_SALTED`, which use just 1 iteration of hashing. Does this make sense, or we should still return an error for such case?

File: include/rnp/rnp.h [hunk 2]
```diff
@@ -1349,6 +1349,81 @@ RNP_API rnp_result_t rnp_key_is_retired(rnp_key_handle_t key, bool *result);
  **/
 RNP_API rnp_result_t rnp_key_is_locked(rnp_key_handle_t key, bool *result);
 
+/**
+ * @brief Get type of protection, used for secret key data.
+ *
+ * @param key key handle, cannot be NULL and should have secret part (see function
+ *            rnp_key_have_secret()).
+ * @param type on success protection type will be stored here. Cannot be NULL.
```
Comment (antonsviridenko) [inline include/rnp/rnp.h:1357]: When I read this, I thought the buffer should be allocated by the caller. But I see in the source that buffer is allocated using strdup() inside the function. So it should be explicitly mentioned that buffer is allocated by the function and that it should be released by rnp_buffer_destroy().

Also, is there a real reason for duplicating these strings - "None", "Encrypted", etc? They are static string literals, and therefore just pointer to them could be returned instead, and type parameter can be declared as `const char **type`. Is there any need that caller could modify contents of returned buffer later? This optimization could reduce memory usage.

File: include/rnp/rnp.h [hunk 3]
```diff
@@ -1349,6 +1349,81 @@ RNP_API rnp_result_t rnp_key_is_retired(rnp_key_handle_t key, bool *result);
...
+ * @return RNP_SUCCESS on success, or any other value on error
+ */
+RNP_API rnp_result_t rnp_key_get_protection_type(rnp_key_handle_t key, char **type);
+
+/**
+ * @brief Get mode in which secret key data is encrypted.
+ *
+ * @param key key handle, cannot be NULL and should have secret part (see function
+ *            rnp_key_have_secret()).
+ * @param mode on success secret key protection mode name will be stored here. Cannot be NULL.
```
Comment (antonsviridenko) [inline include/rnp/rnp.h:1377]: same here

File: include/rnp/rnp.h [hunk 4]
```diff
@@ -1349,6 +1349,81 @@ RNP_API rnp_result_t rnp_key_is_retired(rnp_key_handle_t key, bool *result);
...
+ */
+RNP_API rnp_result_t rnp_key_get_protection_mode(rnp_key_handle_t key, char **mode);
+
+/**
+ * @brief Get cipher, used to encrypt secret key data.
+ *        Note: this call will return an error if secret key data is not available or secret
+ *        key is not encrypted.
+ *
+ * @param key key handle, cannot be NULL and should have secret part.
+ * @param cipher on success cipher name will be stored here. See
```
Comment (antonsviridenko) [inline include/rnp/rnp.h:1397]: see previous comments

File: src/lib/rnp.cpp (hunk 1 above)
Comment (dewyatt) [reply src/lib/rnp.cpp:6299]: @ni4 I thought I responded, but I think I missed the call to `pgp_key_has_encryption_info` and was thinking about the case of `PGP_S2KU_NONE`, so this is ok as-is.

File: include/rnp/rnp.h (hunk 2 above)
Comment (ni4) [reply include/rnp/rnp.h:1357]: @antonsviridenko Thanks for pointing at this, forgot to mention that string should be deallocated. Will fix it.

In all other function calls string constants (like algorithm names) are also duplicated. As I understand, the idea behind this is that caller shouldn't have any direct interpretable pointers to library internal things - all returned values are owned by caller.

**Comment by ni4** [comment]
@antonsviridenko Please see the updated PR.

**Comment by ronaldtse** [comment]
@dewyatt if approved please feel free to merge, thanks!

**Pull request merged**